core/pkg/monitor/openmetrics.go

```diff
@@ -0,0 +1,351 @@
+			processedFilters = processSequenceFilters(v.Sequence)
+		case *spb.OpenMetricsFilters_Mapping:
+			processedFilters = processMappingFilters(v.Mapping)
+		default:
+			logger.Warn("Unknown filter type, using empty filter")
+		}
+	}
+
+	// cache to use with ShouldCaptureMetric
+	cache, err := lru.New(100)
+	if err != nil {
+		logger.Error("openmetrics: error creating cache", "error", err)
+		return nil
+	}
+
+	om := &OpenMetrics{
+		name:        name,
+		url:         url,
+		filters:     processedFilters,
+		client:      client,
+		logger:      logger,
+		labelMap:    make(map[string]map[string]int),
+		labelHashes: make(map[string]map[string]string),
+		cache:       cache,
+	}
+
+	return om
+}
+
+func (o *OpenMetrics) Name() string { return o.name }
+
+func (o *OpenMetrics) SetFilters(filters []Filter) {
+	o.filters = filters
+}
+
+func (o *OpenMetrics) Cache() *lru.Cache {
+	return o.cache
+}
+
+// ShouldCaptureMetric checks if a metric should be captured based on the filters.
+func (o *OpenMetrics) ShouldCaptureMetric(metricName string, metricLabels map[string]string) bool {
+	if len(o.filters) == 0 {
+		return true // If no filters, capture all metrics
+	}
+
+	// generate a hash of metricName and metricLabels to avoid recomputing it
+	// for the same metric
+	hash := o.GenerateLabelHash(metricLabels)
+	if shouldCapture, ok := o.cache.Get(metricName + hash); ok {
+		return shouldCapture.(bool)
+	}
+
+	var shouldCapture bool
+
+	fullMetricName := fmt.Sprintf("%s.%s", o.Name(), metricName)
+
+	for _, filter := range o.filters {
+		if match, _ := regexp.MatchString(filter.MetricNameRegex, fullMetricName); !match {
+			continue
+		}
+
+		// if only metric name regex is provided, capture the metric.
+		if len(filter.LabelFilters) == 0 {
+			o.cache.Add(metricName+hash, true)
+			return true
+		}
+
+		shouldCapture = true
+		for _, labelFilter := range filter.LabelFilters {
+			if labelValue, ok := metricLabels[labelFilter.LabelName]; !ok {
+				shouldCapture = false
+				break
+			} else {
+				if match, _ := regexp.MatchString(labelFilter.LabelRegex, labelValue); !match {
+					shouldCapture = false
+					break
+				}
+			}
+		}
+
+		o.cache.Add(metricName+hash, shouldCapture)
+
+		if shouldCapture {
+			return true
+		}
+	}
+
+	o.cache.Add(metricName+hash, false)
+
+	return false
+}
+
+// Sample fetches and processes metrics from the OpenMetrics endpoint.
+func (o *OpenMetrics) Sample() (map[string]any, error) {
+	resp, err := o.client.Get(o.url)
+	if err != nil {
+		return nil, err
+	}
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+
+	if resp == nil {
+		return nil, fmt.Errorf("could not fetch metrics from endpoint")
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	var parser expfmt.TextParser
+	metricFamilies, err := parser.TextToMetricFamilies(resp.Body)
+	if err != nil {
+		o.logger.Error("Error parsing metrics", "error", err)
```

Comment:
could you add a prefix here for the function?

Reply:
can you explain please?

Reply:
Oh just the regular format we use in other messages;
function name: error-message

Reply:
ah, got it, sry